File: front-end/src/domain/valueobject/location/route-sequence.ts
/**
 * RouteSequence
 *
 * Represents the sequence number of a stop in a route.
 * Immutable value object that encapsulates the sequence position.
 * Used to track the order of stops in a daily plan route.
 */
export class RouteSequence {
  /**
   * Sequence value (order position in route).
   * Must be a positive integer (>= 1).
   */
  private readonly sequence: number;

  /**
   * Private constructor. Use static factory `of` instead.
   * @param sequence The sequence number (must be >= 1)
   */
  private constructor(sequence: number) {
    this.validate(sequence);
    this.sequence = sequence;
  }

  /**
   * Validate that sequence is a positive integer.
   * @param sequence The sequence value to validate
   * @throws Error if sequence is less than 1
   */
  private validate(sequence: number): void {
    if (!Number.isInteger(sequence) || sequence < 1) {
      throw new Error('Route sequence must be a positive integer (>= 1)');
    }
  }

  /**
   * Create a RouteSequence from a sequence number.
   * @param sequence Sequence number (>= 1)
   * @returns A new RouteSequence instance
   * @throws Error if sequence is invalid
   */
  static of(sequence: number): RouteSequence {
    return new RouteSequence(sequence);
  }

  /**
   * Get the sequence value.
   * @returns The sequence number
   */
  getSequence(): number {
    return this.sequence;
  }

  /**
   * Get numeric sequence value.
   * @returns The sequence value
   */
  getValue(): number {
    return this.sequence;
  }

  /**
   * Create a new instance with an updated sequence.
   * @param newSequence The new sequence value
   * @returns A new RouteSequence instance
   * @throws Error if newSequence is invalid
   */
  setValue(newSequence: number): RouteSequence {
    return RouteSequence.of(newSequence);
  }

  /**
   * Get the next sequence in order.
   * @returns A new RouteSequence with sequence + 1
   */
  getNext(): RouteSequence {
    return RouteSequence.of(this.sequence + 1);
  }

  /**
   * Get the previous sequence in order.
   * @returns A new RouteSequence with sequence - 1
   * @throws Error if sequence is 1 (no previous)
   */
  getPrevious(): RouteSequence {
    if (this.sequence === 1) {
      throw new Error('Cannot get previous sequence from sequence 1');
    }
    return RouteSequence.of(this.sequence - 1);
  }

  /**
   * Compare if this sequence is greater than another.
   * @param other The other RouteSequence to compare
   * @returns true if this sequence is greater, false otherwise
   * @throws Error if other is null or undefined
   */
  greaterThan(other: RouteSequence): boolean {
    if (!other) {
      throw new Error('Other RouteSequence cannot be null');
    }
    return this.sequence > other.sequence;
  }

  /**
   * Compare if this sequence is greater than or equal to another.
   * @param other The other RouteSequence to compare
   * @returns true if this sequence is greater or equal, false otherwise
   * @throws Error if other is null or undefined
   */
  greaterThanOrEqual(other: RouteSequence): boolean {
    if (!other) {
      throw new Error('Other RouteSequence cannot be null');
    }
    return this.sequence >= other.sequence;
  }

  /**
   * Compare if this sequence is less than another.
   * @param other The other RouteSequence to compare
   * @returns true if this sequence is less, false otherwise
   * @throws Error if other is null or undefined
   */
  lessThan(other: RouteSequence): boolean {
    if (!other) {
      throw new Error('Other RouteSequence cannot be null');
    }
    return this.sequence < other.sequence;
  }

  /**
   * Compare equality with another object.
   * @param otherObject The object to compare with
   * @returns true if the objects are equal, false otherwise
   */
  equals(otherObject: unknown): boolean {
    if (this === otherObject) {
      return true;
    }
    if (!(otherObject instanceof RouteSequence)) {
      return false;
    }
    return this.sequence === otherObject.sequence;
  }

  /**
   * Get the hash code for this object.
   * @returns A numeric hash code
   */
  hashCode(): number {
    return this.sequence;
  }

  /**
   * Human-readable representation.
   * @returns A formatted string representation
   */
  toString(): string {
    return `RouteSequence={sequence=${this.sequence}}`;
  }
}

File: back-end/src/main/java/es/ull/project/domain/valueobject/route/RouteSequence.java
package es.ull.project.domain.valueobject.route;

import java.util.Objects;

/**
 * RouteSequence
 *
 * Represents the 1-based sequence order of a stop within a route.
 * This is an immutable value object.
 */
public final class RouteSequence {

    private static final String ERROR_SEQUENCE_NON_POSITIVE = "Sequence must be a positive integer greater than zero";
    private static final int MIN_SEQUENCE = 1;

    /**
     * The sequence number.
     * It is a required attribute.
     */
    private final int value;

    /**
     * Private constructor for the value object.
     *
     * @param value The sequence number.
     */
    private RouteSequence(int value) {
        validateSequence(value);
        this.value = value;
    }

    /**
     * Validates that the provided sequence is strictly positive.
     *
     * @param value Sequence to validate.
     * @throws IllegalArgumentException if the sequence is less than 1.
     */
    private void validateSequence(int value) {
        if (value < MIN_SEQUENCE) {
            throw new IllegalArgumentException(ERROR_SEQUENCE_NON_POSITIVE);
        }
    }

    /**
     * Factory method to create an instance from an integer value.
     *
     * @param value The sequence integer.
     * @return A new instance of RouteSequence.
     */
    public static RouteSequence of(int value) {
        return new RouteSequence(value);
    }

    /**
     * Returns the sequence integer value.
     *
     * @return The sequence number.
     */
    public int getValue() {
        return value;
    }

    /**
     * Compares this RouteSequence to another object for equality.
     *
     * @param otherObject the object to compare with
     * @return {@code true} if the other object is a RouteSequence with the same sequence value
     */
    @Override
    public boolean equals(Object otherObject) {
        if (this == otherObject) {
            return true;
        }
        if (otherObject == null || getClass() != otherObject.getClass()) {
            return false;
        }
        RouteSequence other = (RouteSequence) otherObject;
        return value == other.value;
    }

    /**
     * Returns a hash code for this RouteSequence based on its sequence value.
     *
     * @return the hash code
     */
    @Override
    public int hashCode() {
        return Objects.hash(value);
    }

    /**
     * Returns a human-readable string representation of this RouteSequence.
     *
     * @return a string in the format {@code RouteSequence={n}}
     */
    @Override
    public String toString() {
        return String.format("RouteSequence={%d}", value);
    }
}
